pkg/kubelet/cm/dra/plugin/plugin.go

```diff
@@ -132,29 +122,17 @@ func (p *Plugin) NodePrepareResources(
 	defer cancel()
 
 	var response *drapbv1beta1.NodePrepareResourcesResponse
-	switch supportedAPI {
-	case apiV1beta1:
+	switch p.chosenService {
+	case drapbv1beta1.DRAPluginService:
 		nodeClient := drapbv1beta1.NewDRAPluginClient(conn)
 		response, err = nodeClient.NodePrepareResources(ctx, req)
-	case apiV1alpha4:
+	case drapbv1alpha4.NodeService:
 		nodeClient := drapbv1alpha4.NewNodeClient(conn)
 		response, err = nodeClient.NodePrepareResources(ctx, req)
 	default:
-		// Try it, fall back if necessary.
-		supportedAPI = apiV1beta1
-		nodeClient := drapbv1beta1.NewDRAPluginClient(conn)
-		response, err = nodeClient.NodePrepareResources(ctx, req)
-		if err != nil && status.Convert(err).Code() == codes.Unimplemented {
-			supportedAPI = apiV1alpha4
-			nodeClient := drapbv1alpha4.NewNodeClient(conn)
-			response, err = nodeClient.NodePrepareResources(ctx, req)
-		}
-		if err == nil || status.Convert(err).Code() != codes.Unimplemented {
-			// Store discovered version for future use.
-			p.mutex.Lock()
-			p.supportedAPI = supportedAPI
-			p.mutex.Unlock()
-		}
+		// Shouldn't happen, validateSupportedServices should only
+		// return services we support here.
+		return nil, fmt.Errorf("internal error: unsupported chosen service: %q", p.chosenService)
```

Comment:
is it ok that existing DRA plugins will fail? Let's mention it in release notes

Reply:
They don't fail: they report "1.0.0" because a version was required by 1.31, and then `validateSupportedServices` falls back to using the alpha API exactly to prevent them from failing:

```
	// Fall back to alpha if necessary because
	// plugins at that time didn't advertise gRPC services.
	if chosenService == "" {
		chosenService = drapbv1alpha4.NodeService
	}

	return chosenService, nil
```

Reply:
To verify, I ran the DRA E2E tests from Kubernetes 1.31  (minus classic DRA, minus one slow test)  against a cluster using this PR where the resource.k8s.io/v1alpha3 is enabled. This covers version skew between a client (e2e.test) and the apiserver as well as version skew between a DRA driver (the test driver from 1.31) and kubelet. All passed.

I then replaced kubelet with kubelet from 1.31 and ran the E2E tests from this PR. This covers version skew between kubelet and the control plane. However, that fails because drivers now don't provide a semantic version in the "supported versions", which causes kubelet from 1.31 to reject registation:
  
```
              {
                  FullMethod: "/pluginregistration.Registration/NotifyRegistrationStatus",
                  Request: 
                      error: 'RegisterPlugin error -- plugin validation failed with err: invalid versions
                        of plugin dra-8953.k8s.io: none of the versions are supported: could not find a
                        highest supported version from versions ([v1beta1.DRAPlugin]) reported: illegal
                        version string "v1beta1.DRAPlugin"',
                  Response: 
                      null,
                  Err: <*errors.errorString | 0xc000e10a90>{
                      s: "failed registration process: RegisterPlugin error -- plugin validation failed with err: invalid versions of plugin dra-8953.k8s.io: none of the versions are supported: could not find a highest supported version from versions ([v1beta1.DRAPlugin]) reported: illegal version string \"v1beta1.DRAPlugin\"",
                  },
              },
```

We could tweak the helper code a bit: if a DRA driver is configured to *only* enable the alpha gRPC server, then we can report "1.0.0" instead of the (according to the new rules) correct "v1alpha3.Node". Then such a driver works in this version skew scenario. Whether any DRA vendor will *support* that is a different question, but it seems useful to have the ability. I added it in a separate commit ("DRA driver: support kubelet 1.31").

More important for DRA in general is that version skew between kubelet and the control plane works as intended, which will become important for future Kubernetes releases.